### PR TITLE
Compatible target notifications

### PR DESCRIPTION
**Pull request by jdrosales17**
## TASKS DONE

- [X] Add compatible target logic to model
- [X] Create device model to store user devices
- [X] Add device create route and action in controller
- [X] Add one signal service to send push notifications
- [X] Update create action in targets controller to support push notifications when a compatible target is created
- [X] Update create request test

### CODE REVIEW

File: app/models/target.rb [hunk 1]
```diff
@@ -19,24 +19,38 @@
...
+                   lng_column_name: :longitude
+
   validates :title, presence: true, uniqueness: true
   validates :latitude, :longitude, presence: true
 
   validate :target_limit_per_user
 
+  def search_compatible_targets
+    compatible_targets = []
+    Target.where.not(user_id: user_id).where(topic_id: topic_id).each do |target|
```
Comment (Alxzu) [inline app/models/target.rb]: I'll suggest creating two scopes on Target like `scope :all_except_mine, ->(user_id) { where.not(user_id: user_id) }` & `scope :with_topic, (topic_id) { where(topic_id: topic_id) }` and then call it here like `Target.all_except_mine(user_id).with_topic(topic_id)`

File: spec/requests/api/v1/users/update_spec.rb [hunk 2]
```diff
@@ -2,8 +2,8 @@
 
 # Test suite for PUT /api/v1/user/:id
 describe 'PUT /api/v1/user/:id', type: :request do
-  let(:user)    { create(:user) }
-  let(:user2)   { create(:user) }
+  let(:user)  { create(:user) }
+  let(:user2) { create(:user) }
```
Comment (Alxzu) [inline spec/requests/api/v1/users/update_spec.rb]: Maybe change `user2` to `other_user`

File: app/controllers/api/v1/targets_controller.rb [hunk 3]
```diff
@@ -11,6 +11,15 @@ def index
       # POST /api/v1/targets
       def create
         @target = current_user.targets.create!(target_params)
```
Comment (GuilleLeopold) [inline app/controllers/api/v1/targets_controller.rb:13]: I know it's outside the scope of this PR but target should be a helper method

File: app/controllers/api/v1/targets_controller.rb [hunk 4]
```diff
@@ -11,6 +11,15 @@ def index
...
+        @compatible_users = []
+        @target.search_compatible_targets.each do |compatible_target|
```
Comment (GuilleLeopold) [inline app/controllers/api/v1/targets_controller.rb]: You can change this a bit doing something like this: 
```
@compatible_users = target.map do |compatible_target|
   compatible_target_user = compatible_target.user
   begin 
          notify(...)
   rescue RuntimeError => e
          logger.error e
   end
   compatible_target_user
end
```

File: app/models/target.rb (hunk 1 above)
Comment (GuilleLeopold) [reply app/models/target.rb]: also as you are already in the Target model, no need to call `Target`

Comment (GuilleLeopold) [reply app/models/target.rb]: You can apply the same logic as in my previous comment, using map

File: app/services/one_signal_service.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class OneSignalService
```
Comment (GuilleLeopold) [inline app/services/one_signal_service.rb]: As you don't need to have an instance of this class, you can change this into a `module` and extend self to declare the class methods (if you have any doubts, it's specified in the rootstrap rails guides how to do this). This way when you call the OneSignal  service you wont need to use `.new`

File: app/services/one_signal_service.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  include HTTParty
+
+  base_uri 'https://onesignal.com/api/v1'
```
Comment (GuilleLeopold) [inline app/services/one_signal_service.rb]: Add a comment  explaining how this work

File: app/services/one_signal_service.rb (hunk 5 above)
Comment (GuilleLeopold) [reply app/services/one_signal_service.rb]: Also, there is a OneSignal gem available, anyway you can keep this. It's just a comment to let you know its existance

File: spec/requests/api/v1/passwords/create_spec.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  let(:user) { create(:user) }
+
+  before(:each) { user.confirm }
+
+  context 'when the request is valid' do
+    before do
+      post  '/api/v1/auth/password',
+            params: {
+              email: user.email,
+              redirect_url: 'https://www.google.com'
```
Comment (GuilleLeopold) [inline spec/requests/api/v1/passwords/create_spec.rb]: I would use a random url here 🤔 Anyway I think we already discuss this and maybe there is a rebase missing or something like this

**Review by Alxzu — APPROVED** [review]

**Review by GuilleLeopold — APPROVED** [review]

**Pull request merged**